qiskit/utils/deprecation.py

```diff
@@ -113,3 +113,117 @@ def _rename_kwargs(
                 )
 
                 kwargs[new_arg] = kwargs.pop(old_arg)
+
+
+# We insert deprecations in-between the description and Napoleon's meta sections. The below is from
+# https://www.sphinx-doc.org/en/master/usage/extensions/napoleon.html#docstring-sections. We use
+# lowercase because Napoleon is case-insensitive.
+_NAPOLEON_META_LINES = frozenset(
+    {
+        "args:",
+        "arguments:",
+        "attention:",
+        "attributes:",
+        "caution:",
+        "danger:",
+        "error:",
+        "example:",
+        "examples:",
+        "hint:",
+        "important:",
+        "keyword args:",
+        "keyword arguments:",
+        "note:",
+        "notes:",
+        "other parameters:",
+        "parameters:",
+        "return:",
+        "returns:",
+        "raises:",
+        "references:",
+        "see also:",
+        "tip:",
+        "todo:",
+        "warning:",
+        "warnings:",
+        "warn:",
+        "warns:",
+        "yield:",
+        "yields:",
+    }
+)
+
+
+def add_deprecation_to_docstring(
+    func: Callable, msg: str, *, since: Optional[str], pending: bool
+) -> None:
+    """Dynamically insert the deprecation message into ``func``'s docstring.
+
+    Args:
+        func: The function to modify.
+        msg: The full deprecation message.
+        since: The version the deprecation started at.
+        pending: Is the deprecation still pending?
+    """
+    if "\n" in msg:
+        raise ValueError(
+            "Deprecation messages cannot contain new lines (`\\n`), but the deprecation for "
+            f'{func.__qualname__} had them. Usually this happens when using `"""` multiline '
+            f"strings; instead, use string concatenation.\n\n"
+            "This is a simplification to facilitate deprecation messages being added to the "
+            "documentation. If you have a compelling reason to need "
+            "new lines, feel free to improve this function or open a request at "
+            "https://github.com/Qiskit/qiskit-terra/issues."
+        )
```

Comment:
It feels a bit hypocritical to complain that the developer uses a newline in their warning message when we do the exact same thing in this error message ;).